### ISSUE TITLE
vips 8.15.1 build segfaulting with clang

### ISSUE THREAD

**Issue by der-eismann** [libvips/libvips#3874]
## Bug report

**Describe the bug**
We are using vips as a dependency in another project and compiled it by ourselves, to squeeze out some performance. While clang was no issue until 8.14, it suddenly started segfaulting with 8.15. To be more precise it's faulting while linking, however using mold or the default linker doesn't make a difference and I'm not sure how to get verbose output from it as it says `use -v to see invocation`, but where should I use it?
I'm sorry if it's caused by our setup, but it has worked so far and I'm a bit puzzled where the segfault is coming from.

**To Reproduce**
Build the following image, it will fail. When switching `clang` to `gcc` and `clang++` to `g++`, it will succeed.
```Dockerfile
FROM registry.fedoraproject.org/fedora:39

ENV LIBVIPS_VERSION=8.15.1

ENV CC="clang"
ENV CXX="clang++"
ENV CFLAGS="-flto -I/opt/vips/include -O3 -march=x86-64-v3 -maes -mfsgsbase -mxsaveopt -mtune=alderlake"
ENV CXXFLAGS=${CFLAGS}
ENV LDFLAGS="-fuse-ld=mold -Wl,-rpath,/opt/vips/lib -L/opt/vips/lib"
ENV PKG_CONFIG_PATH="/opt/vips/lib/pkgconfig"
ENV LD_LIBRARY_PATH="/opt/vips/lib"

RUN dnf install -y https://mirrors.rpmfusion.org/free/fedora/rpmfusion-free-release-$(rpm -E %fedora).noarch.rpm && \
	dnf install -y clang cmake mold make diffutils findutils file git git-lfs xz \
		glib2-devel expat-devel libjpeg-turbo-devel fftw-devel orc-devel libwebp-devel libpng-devel \
		libimagequant-devel libde265-devel libtiff-devel libheif-devel libcgif-devel librsvg2-devel rav1e-devel libdav1d-devel \
		meson libexif-devel

RUN curl -L https://github.com/libvips/libvips/releases/download/v${LIBVIPS_VERSION}/vips-${LIBVIPS_VERSION}.tar.xz | tar xJC /tmp --no-same-owner

WORKDIR /tmp/vips-${LIBVIPS_VERSION}

RUN meson setup build --prefix /opt/vips && \
	cd build && \
	meson compile -j 4 && \
	meson test && \
	meson install
```

**Expected behavior**
A successful build

**Actual behavior**
```
…
[481/508] Compiling C++ object fuzz/thumbnail_fuzzer.p/thumbnail_fuzzer.cc.o
[482/508] Compiling C++ object fuzz/smartcrop_fuzzer.p/smartcrop_fuzzer.cc.o
[483/508] Compiling C++ object fuzz/mosaic_fuzzer.p/mosaic_fuzzer.cc.o
[484/508] Linking target libvips/libvips.so.42.17.1
FAILED: libvips/libvips.so.42.17.1 
clang++  -o libvips/libvips.so.42.17.1 libvips/libvips.so.42.17.1.p/meson-generated_.._include_vips_enumtypes.c.o -L/opt/vips/lib -I/opt/vips/include -Wl,--as-needed -Wl,--no-undefined -Wl,-O1 -shared -fPIC -Wl,--start-group -Wl,-soname,libvips.so.42 -Wl,--whole-archive libvips/foreign/libnsgif/libnsgif.a libvips/foreign/libforeign.a libvips/deprecated/libdeprecated.a libvips/arithmetic/libarithmetic.a libvips/resample/libresample.a libvips/colour/libcolour.a libvips/conversion/libconversion.a libvips/convolution/libconvolution.a libvips/freqfilt/libfreqfilt.a libvips/histogram/libhistogram.a libvips/draw/libdraw.a libvips/iofuncs/libiofuncs.a libvips/morphology/libmorphology.a libvips/mosaicing/libmosaicing.a libvips/create/libcreate.a -Wl,--no-whole-archive -fuse-ld=mold -Wl,-rpath,/opt/vips/lib -flto -O3 -march=x86-64-v3 -maes -mfsgsbase -mxsaveopt -mtune=alderlake -Wl,-z,nodelete /usr/lib64/libglib-2.0.so /usr/lib64/libgio-2.0.so /usr/lib64/libgobject-2.0.so /usr/lib64/libgmodule-2.0.so -pthread /usr/lib64/libexpat.so -lm /usr/lib64/libz.so /usr/lib64/libfftw3.so /usr/lib64/libimagequant.so /usr/lib64/libcgif.so /usr/lib64/libexif.so /usr/lib64/libjpeg.so /usr/lib64/libpng16.so /usr/lib64/libwebp.so /usr/lib64/libwebpmux.so /usr/lib64/libwebpdemux.so /usr/lib64/libtiff.so /usr/lib64/librsvg-2.so /usr/lib64/libgdk_pixbuf-2.0.so /usr/lib64/libcairo.so /usr/lib64/libopenjp2.so /usr/lib64/libhwy.so -Wl,--end-group
clang++: error: unable to execute command: Segmentation fault (core dumped)
clang++: error: linker command failed due to signal (use -v to see invocation)
ninja: build stopped: subcommand failed.
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /usr/bin/ninja -j 4
```

**Environment**
(please complete the following information)
 - OS: Fedora 39
 - Vips: 8.15.1

**Comment by jcupitt** [comment]
Hi @der-eismann,

I'd guess this is the C++ stdlib. 

C++ programs need some code (not the `std::` functions, this is more basic) linked in at runtime to support things like memory allocation and exception handling. clang and gcc handle this differently (from memory, clang needs a separate item on the link line? I forget) and you need to be cautious. If you link the wrong libc++, you'll get crashes.

It works OK on ubuntu 23.10, I see:

```
$ CC=clang CXX=clang++ meson setup build --prefix ~/vips --libdir lib -Dmagick=disabled
... stuff
$ cd build/
$ ninja install
...
```

And we build with clang for fuzzing, so it should work.

**Comment by jcupitt** [comment]
It looks like it's failing in library introspection. You probably don't need this, try adding `-Dintrospection=disabled` to your meson setup line.

**Comment by kleisauke** [comment]
This appears to be a crash in Clang during linking with the `-flto` flag; unfortunately, this happens while linking `libvips.so.42`, so `-Dintrospection=disabled` does not mitigate the issue.
<details>
  <summary>Details</summary>

```console
$ coredumpctl debug
[...]
Program terminated with signal SIGSEGV, Segmentation fault.
#0  std::vector<std::unique_ptr<llvm::GlobalValueSummary, std::default_delete<llvm::GlobalValueSummary> >, std::allocator<std::unique_ptr<llvm::GlobalValueSummary, std::default_delete<llvm::GlobalValueSummary> > > >::data (this=0x18)
    at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/stl_vector.h:1263
1263	      { return _M_data_ptr(this->_M_impl._M_start); }
(gdb) bt
#0  std::vector<std::unique_ptr<llvm::GlobalValueSummary, std::default_delete<llvm::GlobalValueSummary> >, std::allocator<std::unique_ptr<llvm::GlobalValueSummary, std::default_delete<llvm::GlobalValueSummary> > > >::data (this=0x18)
    at /usr/bin/../lib/gcc/x86_64-redhat-linux/13/../../../../include/c++/13/bits/stl_vector.h:1263
#1  llvm::ArrayRef<std::unique_ptr<llvm::GlobalValueSummary, std::default_delete<llvm::GlobalValueSummary> > >::ArrayRef<std::allocator<std::unique_ptr<llvm::GlobalValueSummary, std::default_delete<llvm::GlobalValueSummary> > > > (
    Vec=<error reading variable: Cannot access memory at address 0x20>, this=<optimized out>) at /usr/src/debug/llvm-17.0.6-3.fc39.x86_64/include/llvm/ADT/ArrayRef.h:97
#2  llvm::ValueInfo::getSummaryList (this=<optimized out>) at /usr/src/debug/llvm-17.0.6-3.fc39.x86_64/include/llvm/IR/ModuleSummaryIndex.h:189
#3  llvm::computeDeadSymbolsAndUpdateIndirectCalls(llvm::ModuleSummaryIndex&, llvm::DenseSet<unsigned long, llvm::DenseMapInfo<unsigned long, void> > const&, llvm::function_ref<llvm::PrevailingType (unsigned long)>)::$_0::operator()(llvm::ValueInfo, bool) const (this=0x7ffcdd4c04d0, VI=..., IsAliasee=false) at /usr/src/debug/llvm-17.0.6-3.fc39.x86_64/lib/Transforms/IPO/FunctionImport.cpp:962
#4  0x00007f3bbc247bee in llvm::computeDeadSymbolsAndUpdateIndirectCalls(llvm::ModuleSummaryIndex&, llvm::DenseSet<unsigned long, llvm::DenseMapInfo<unsigned long, void> > const&, llvm::function_ref<llvm::PrevailingType (unsigned long)>)
    (Index=..., GUIDPreservedSymbols=..., isPrevailing=...) at /usr/src/debug/llvm-17.0.6-3.fc39.x86_64/lib/Transforms/IPO/FunctionImport.cpp:1013
#5  0x00007f3bbc7e6a72 in llvm::computeDeadSymbolsWithConstProp(llvm::ModuleSummaryIndex&, llvm::DenseSet<unsigned long, llvm::DenseMapInfo<unsigned long, void> > const&, llvm::function_ref<llvm::PrevailingType (unsigned long)>, bool) (
    Index=..., GUIDPreservedSymbols=..., isPrevailing=..., ImportEnabled=<optimized out>) at /usr/src/debug/llvm-17.0.6-3.fc39.x86_64/lib/Transforms/IPO/FunctionImport.cpp:1034
#6  llvm::lto::LTO::run(std::function<llvm::Expected<std::unique_ptr<llvm::CachedFileStream, std::default_delete<llvm::CachedFileStream> > > (unsigned int, llvm::Twine const&)>, std::function<llvm::Expected<std::function<llvm::Expected<std::unique_ptr<llvm::CachedFileStream, std::default_delete<llvm::CachedFileStream> > > (unsigned int, llvm::Twine const&)> > (unsigned int, llvm::StringRef, llvm::Twine const&)>) (this=0x55e9f9a8f370, AddStream=..., Cache=...)
    at /usr/src/debug/llvm-17.0.6-3.fc39.x86_64/lib/LTO/LTO.cpp:1174
#7  0x00007f3bcf2d782b in runLTO () at /usr/src/debug/llvm-17.0.6-3.fc39.x86_64/tools/gold/gold-plugin.cpp:1099
#8  allSymbolsReadHook () at /usr/src/debug/llvm-17.0.6-3.fc39.x86_64/tools/gold/gold-plugin.cpp:1123
#9  0x00007f3bcf2d1b3e in all_symbols_read_hook () at /usr/src/debug/llvm-17.0.6-3.fc39.x86_64/tools/gold/gold-plugin.cpp:1148
#10 0x000055e9f69a57b2 in plugin_call_all_symbols_read () at ../../ld/plugin.c:1329
#11 lang_process () at ../../ld/ldlang.c:7943
#12 0x000055e9f698e533 in main (argc=96, argv=<optimized out>) at ../../ld/ldmain.c:499
```
</details>

A similiar crash was reported in https://github.com/llvm/llvm-project/issues/62613.

**Comment by jcupitt** [comment]
I think the `cc.links` fix resolves this, and will be in 8.15.2. I'll close.

**Comment by kleisauke** [comment]
This appears to impact the build of the OSS-Fuzz introspector as well, see:
https://oss-fuzz-build-logs.storage.googleapis.com/log-28bb0e48-641c-430c-b5d3-45963d09534f.txt
https://github.com/google/oss-fuzz/blob/bf577d8664b7f7c06ba42f95c01f9161c73cd5ff/infra/base-images/base-builder/Dockerfile#L73C39-L73C44

<details>
  <summary>Details</summary>

```console
$ gdb /usr/bin/x86_64-linux-gnu-ld.gold 'core.ld\x2egold.1000.f0e7e2cf4aaa4465acc8476884e70e1f.217441.1714902089000000'   
[...]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Core was generated by `/usr/bin/ld.gold -pie -z relro --hash-style=gnu --eh-frame-hdr -m elf_x86_64 -d'.
Program terminated with signal SIGSEGV, Segmentation fault.

#0  0x00007fba4c73e936 in llvm::computeDeadSymbolsAndUpdateIndirectCalls(llvm::ModuleSummaryIndex&, llvm::DenseSet<unsigned long, llvm::DenseMapInfo<unsigned long, void> > const&, llvm::function_ref<llvm::PrevailingType (unsigned long)>)::$_0::operator()(llvm::ValueInfo, bool) const () from /usr/local/bin/../lib/LLVMgold.so
(gdb) bt
#0  0x00007fba4c73e936 in llvm::computeDeadSymbolsAndUpdateIndirectCalls(llvm::ModuleSummaryIndex&, llvm::DenseSet<unsigned long, llvm::DenseMapInfo<unsigned long, void> > const&, llvm::function_ref<llvm::PrevailingType (unsigned long)>)::$_0::operator()(llvm::ValueInfo, bool) const () from /usr/local/bin/../lib/LLVMgold.so
#1  0x00007fba4c73e7bd in llvm::computeDeadSymbolsAndUpdateIndirectCalls(llvm::ModuleSummaryIndex&, llvm::DenseSet<unsigned long, llvm::DenseMapInfo<unsigned long, void> > const&, llvm::function_ref<llvm::PrevailingType (unsigned long)>)
    () from /usr/local/bin/../lib/LLVMgold.so
#2  0x00007fba4c73eb62 in llvm::computeDeadSymbolsWithConstProp(llvm::ModuleSummaryIndex&, llvm::DenseSet<unsigned long, llvm::DenseMapInfo<unsigned long, void> > const&, llvm::function_ref<llvm::PrevailingType (unsigned long)>, bool) ()
   from /usr/local/bin/../lib/LLVMgold.so
#3  0x00007fba4c6ded9e in llvm::lto::LTO::run(std::function<llvm::Expected<std::unique_ptr<llvm::CachedFileStream, std::default_delete<llvm::CachedFileStream> > > (unsigned int, llvm::Twine const&)>, std::function<llvm::Expected<std::function<llvm::Expected<std::unique_ptr<llvm::CachedFileStream, std::default_delete<llvm::CachedFileStream> > > (unsigned int, llvm::Twine const&)> > (unsigned int, llvm::StringRef, llvm::Twine const&)>) ()
   from /usr/local/bin/../lib/LLVMgold.so
#4  0x00007fba4c3e4210 in all_symbols_read_hook() () from /usr/local/bin/../lib/LLVMgold.so
#5  0x000055b7e853c28f in ?? ()
#6  0x000055b7e853c3e0 in ?? ()
#7  0x000055b7e8581b98 in ?? ()
#8  0x000055b7e8581dea in ?? ()
#9  0x000055b7e8451ee5 in ?? ()
#10 0x00007fba4e9c5083 in __libc_start_main (main=0x55b7e84518d0, argc=103, argv=0x7ffe2a4eb388, init=<optimized out>, fini=<optimized out>, rtld_fini=<optimized out>, stack_end=0x7ffe2a4eb378) at ../csu/libc-start.c:308
#11 0x000055b7e8452ace in ?? ()
```
</details>

I haven't investigated it further yet, but I can confirm that building without `-flto` resolves this (see PR #3952).

> There's a downstream Gentoo bug about this too - see https://bugs.gentoo.org/921728

According to that bug report, this issue is not related to Highway.

**Comment by kleisauke** [comment]
I just opened PR #3953 for this and reported this reduced example upstream at https://github.com/llvm/llvm-project/issues/70184#issuecomment-2094786264.
